test/e2e/federation-authn.go

```diff
@@ -0,0 +1,110 @@
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package e2e
+
+import (
+	"fmt"
+
+	"k8s.io/kubernetes/federation/client/clientset_generated/federation_release_1_4"
+	"k8s.io/kubernetes/pkg/api/errors"
+	"k8s.io/kubernetes/pkg/client/unversioned/clientcmd"
+	clientcmdapi "k8s.io/kubernetes/pkg/client/unversioned/clientcmd/api"
+	"k8s.io/kubernetes/test/e2e/framework"
+
+	. "github.com/onsi/ginkgo"
+	. "github.com/onsi/gomega"
+)
+
+var _ = framework.KubeDescribe("[Feature:Federation]", func() {
+	f := framework.NewDefaultFederatedFramework("federation-apiserver-authn")
+
+	var _ = Describe("Federation API server authentication", func() {
+		BeforeEach(func() {
+			framework.SkipUnlessFederated(f.Client)
+		})
+
+		It("should accept cluster resources when the client has right authentication credentials", func() {
+			framework.SkipUnlessFederated(f.Client)
+
+			svc := createServiceOrFail(f.FederationClientset_1_4, f.Namespace.Name)
+			deleteServiceOrFail(f.FederationClientset_1_4, f.Namespace.Name, svc.Name)
+		})
+
+		It("should not accept cluster resources when the client has invalid authentication credentials", func() {
+			framework.SkipUnlessFederated(f.Client)
+
+			contexts := f.GetUnderlyingFederatedContexts()
+
+			fcs, err := invalidAuthFederationClientSet(contexts[0].User)
+			framework.ExpectNoError(err)
+
+			svc, err := createService(fcs, f.Namespace.Name)
+			Expect(errors.IsUnauthorized(err)).To(BeTrue())
+			if err == nil && svc != nil {
+				deleteServiceOrFail(fcs, f.Namespace.Name, svc.Name)
+			}
+		})
+
+		It("should not accept cluster resources when the client has no authentication credentials", func() {
+			framework.SkipUnlessFederated(f.Client)
+
+			fcs, err := invalidAuthFederationClientSet(nil)
+			ExpectNoError(err)
+
+			svc, err := createService(fcs, f.Namespace.Name)
+			Expect(errors.IsUnauthorized(err)).To(BeTrue())
+			if err == nil && svc != nil {
+				deleteServiceOrFail(fcs, f.Namespace.Name, svc.Name)
+			}
+		})
+	})
+})
+
+func invalidAuthFederationClientSet(user *framework.KubeUser) (*federation_release_1_4.Clientset, error) {
+	overrides := &clientcmd.ConfigOverrides{}
+	if user != nil {
+		overrides = &clientcmd.ConfigOverrides{
+			AuthInfo: clientcmdapi.AuthInfo{
+				Token:    user.User.Token,
+				Username: user.User.Username,
+				Password: user.User.Password,
+			},
+		}
+	}
+
+	config, err := framework.LoadFederatedConfig(overrides)
+	if err != nil {
+		return nil, err
+	}
+
+	if user == nil {
+		config.Password = ""
+		config.BearerToken = ""
+		config.Username = ""
+	}
+
+	c, err := federation_release_1_4.NewForConfig(config)
```

Comment:
@nikhiljindal there is another reason I did not refactor `LoadFederationClientset_1_4()` which I forgot to mention. I need to explicitly set `config.{Password, BearerToken, Username}` to empty strings for one of the tests. It is too cumbersome to do it after the `ClientSet` is created because every client in the set has a copy of the config. If I have to set these values after the `ClientSet` is created, I have to update all of them, which is a lot more code and less neater than this.

If you insist I will be happy to export `setTimeouts` function and re-use it.